### ISSUE TITLE
[FSDP] How to use customized backward hooks?

### ISSUE THREAD

**Issue by woodyx218** [facebookresearch/fairscale#1101]
Hi, my use case needs a module backward hook, which computes a scaled gradient (saved to the attribute param.scaled_grad). At optimizer step, I replace param.grad=param.scaled_grad, and then the model is updated with my scaled gradient. While this works fine with single GPU, FSDP will not find my scaled gradient (i.e. hasattr(param,'scaled_grad')==False) in the FSDP._post_backward_hook.

My understanding is that somehow the parameter in self.params is not the one that has the scaled gradient, because it is flattened and sharded. Given that param.grad can be accessed in FSDP._post_backward_hook, maybe a slight modification can allow me to also access other attributes that are instantiated during the module backward hook.

Please let me know if an example is needed. BTW, I am using torch==1.11.

**Comment by min-xu-ai** [comment]
Interesting use case. Thanks for sharing it.

It kind of depends on whether or not your backward hook runs before FSDP's backward hook, right? If your hook runs first in the backward pass, then param.scaled_grad should be there, right?

As to flatten and sharding, the FSDP backward hook does the sharding for the grad. The grad into the hook should have the original flattened size but then it should be sharded to save memory space. The original params are replaced with views into a flat param and the grad is available on that flat param when all grad are computed on that flat param.

Does the above help a bit?

**Comment by woodyx218** [comment]
Thanks @min-xu-ai for the prompt response. I think by design, my backward hook (a module hook) is called before the gradient is instantiated, and the FSDP backward hook (a tensor hook) is called immediately after the gradient is created, right? However, the param.scaled_grad is not there.

My printout shows that my backward hook is working but the scaled grad is somehow lost, so that replacing the original grad with my scaled grad is not possible. I can provide a code snippet tomorrow.

**Comment by min-xu-ai** [comment]
what's the shape of your scaled_grad? Is it on the flatten param or the original? If it is lost, maybe this is a question for pytorch folks? Maybe you can try to test if the same thing happens with multiple backward hooks without FSDP?

**Comment by woodyx218** [comment]
The scaled_grad is 2D shaped on the original param. I think the problem is how to transfer it to the flattened param. I can think of a flowchart, but not sure how can I implement step 3: 

1. my module backward hook instantiate the param.scaled_grad on the original param;
2. the original param's grad is computed and saved to param.grad,
3. on original param, replace param.grad=param.scaled_grad
4. FSDP flatten and shard param.grad
Any advice is appreciated.

**Comment by min-xu-ai** [comment]
I am a bit confused. The flattening does not happen in the backward pass. It is done at the init time. So in your backward pass you should see a 1d tensor. 

Maybe you can try disabling flattening?

**Comment by woodyx218** [comment]
Thanks again for the reply. Here is a code snippet based on FSDP MNIST example (https://drive.google.com/file/d/1LS2s3IR8dt7-NG6KPURfMpObafnDtkjz/view?usp=share_link). My customized backward hook is added between Line 121 and Line 126.

As you can see, the backward is working and param.scaled_grad is instantiated as a 2D tensor. I wonder whether the next step is that the original param.grad is instantiated as 2D tensor then flattened to 1D, or directly instantiated as 1D tensor?

**Comment by min-xu-ai** [comment]
I don't have access to the file. can you paste the code here maybe?

**Comment by woodyx218** [comment]
Sorry. I have allowed access to everyone now.

**Comment by min-xu-ai** [comment]
I see. In the backward hook, you directly access the param:
```
        module.weight.scaled_grad=torch.ones_like(module.weight)
        print('customized hook is working')
```

But I thought backward hook actually passes you the grad. Did you check the grad there to make sure they are what you expected?

Directly access module.weight is not going to access the underlying nn.Parameter object. module.weight is replaced with a view into the underlying flatten parameter in that case. That might explain why you can't see the scaled_grad attr.

**Comment by woodyx218** [comment]
I think the module backward hook is triggered before the grad is created. This can be checked by printing hasattr(param,'grad'). That's why I put the grad in step 2 after step 1 (https://github.com/facebookresearch/fairscale/issues/1101#issuecomment-1381243855).

I am not very familiar with the "view" part. Do you mean that the grad is directly created in a flattened 1D shape, not firstly created in 2D (like in the single GPU case) and then flattened to 1D? In that case, could you suggest a way to replace that 1D grad with my 2D scaled_grad? Many thanks in advance.

**Comment by min-xu-ai** [comment]
> I think the module backward hook is triggered before the grad is created. This can be checked by printing hasattr(param,'grad'). That's why I put the grad in step 2 after step 1 ([#1101 (comment)](https://github.com/facebookresearch/fairscale/issues/1101#issuecomment-1381243855)).

I mean doesn't the callback itself has an input param that is the newly computed grad. I don't mean param.grad.

> I am not very familiar with the "view" part. Do you mean that the grad is directly created in a flattened 1D shape, not firstly created in 2D (like in the single GPU case) and then flattened to 1D? In that case, could you suggest a way to replace that 1D grad with my 2D scaled_grad? Many thanks in advance.

yes, it is directly created on the 1D tensor. You need to go through complicated logic to "unflatten" it and get back the original 2D tensors (many tensors are combined into a single flatten tensor).

**Comment by woodyx218** [comment]
I see. Could you elaborate why I need to unflatten the original 1D grad? I think all I need is to use that complicated logic to flatten my 2D scaled_grad to 1D (so as to replace the orignal 1D grad before reduce scatter). If you can point me to the complicated logic, maybe I can find a way.

Otherwise, it seems FSDP does not generally support this kind of grad manipulation.

**Comment by min-xu-ai** [comment]
You don’t have to. If your use case doesn’t care about the shape you can just work with the 1d tensor directly. 

The complicated logic is in the flatten param wrapper class in this repo.

**Comment by woodyx218** [comment]
That sounds great! But in the module backward hook printout, param.scaled_grad is 2D. How I can directly work with the 1D scaled_grad tensor? 

I did check the _post_backward_hook function in https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/data_parallel/fully_sharded_data_parallel.py
This is the function called before reduce scatter, but the param has no param.scaled_grad, so I don't know where can I find the 1D scaled_grad if it actually exists.

**Comment by min-xu-ai** [comment]
Doesn’t the 1d grad tensor get passed in your hook function?

**Comment by woodyx218** [comment]
I don't understand. The hook function is triggered before the grad (2D or 1D) is created. So in my hook function, there is no grad tensor.

**Comment by min-xu-ai** [comment]
Check the hook function’s input parameters. I think one of them is the computed grad.

**Comment by min-xu-ai** [comment]
Maybe you can register your hook on the tensor?

**Comment by min-xu-ai** [comment]
You can try flatten_parameters=False. When flatten, I am afraid that the extra attrs are lost.

**Comment by woodyx218** [comment]
Thanks for the suggestion. Can you elaborate where shall I apply this setting? I checked the torch FSDP and there is no argument "flatten_parameters" (https://github.com/pytorch/pytorch/blob/master/torch/distributed/fsdp/fully_sharded_data_parallel.py) but this argument exists in fairscale FSDP (https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/data_parallel/fully_sharded_data_parallel.py). 

Why is there a difference between the two FSDP? And which one is up-to-date?

**Comment by min-xu-ai** [comment]
You are right. Torch version is going to be more production ready if not already. Fairscale version is for research and experimentation hence has more features but not all features work equally well. 

For non flatten you need to use the fairscale version.

**Comment by woodyx218** [comment]
It works! One last question, what is the pros/cons of setting flatten_parameters=True/False?

**Comment by min-xu-ai** [comment]
> It works! One last question, what is the pros/cons of setting flatten_parameters=True/False?

of course. The benefit of flatten is better performance since network communication are done with bigger tensors. If you don't flatten, you can run slower. But that's not always the case, you have to benchmark with your model to know for sure. As you have discovered, non-flatten might be slightly more compatible with the original model code in some cases.

please reopen if you still have more questions.
